build/Shared/TaskResultCache.cs

```diff
@@ -0,0 +1,126 @@
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to use for signaling that an operation should be cancelled.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, Func<Task<TValue>> valueFactory, CancellationToken cancellationToken)
+        {
+            return GetOrAddAsync(key, refresh: false, valueFactory, cancellationToken);
+        }
+
+        /// <summary>
+        /// Gets the cached async operation associated with the specified key, or runs the operation asynchronously and returns <see cref="Task{TValue}" /> that the caller can await, and optionally refreshes the cache.
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory)
+        {
+            return GetOrAddAsync(key, refresh, valueFactory, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Gets the cached async operation associated with the specified key, or runs the operation asynchronously and returns <see cref="Task{TValue}" /> that the caller can await, and optionally refreshes the cache.
+        /// </summary>
+        /// <param name="key">The key for the async operation to get or store in the cache.</param>
+        /// <param name="refresh"><see langword="true" /> to force the specified asynchronous operation to be executed and stored in the cache even if a cached operation exists, otherwise <see langword="false" />.</param>
+        /// <param name="valueFactory">A <see cref="Func{TResult}" /> to execute asynchronously if a cached operation does not exist.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> to use for signaling that an operation should be cancelled.</param>
+        /// <returns>A <see cref="Task{TResult}" /> for the specified asynchronous operation from the cache if found, otherwise the scheduled asynchronous operation to await.</returns>
+        public Task<TValue> GetOrAddAsync(TKey key, bool refresh, Func<Task<TValue>> valueFactory, CancellationToken cancellationToken)
+        {
+            if (!refresh && _cache.TryGetValue(key, out Task<TValue>? result))
```

Comment:
> said it would reduce lock contention if we always look for an existing value and then lock before writing

It will reduce lock contention, but this pattern is not safe with `Dictionary<>`.  You would need to use `ConcurrentDictionary<>` or something similar to avoid locking on the read path. Technically you could also use the non-generic Hashtable, which as an implementation detail is safe for reads concurrently with at most one write.

Reply:
I've pushed a commit to use ConcurrentDictionary but still just call `TryGetValue` then manually lock.  Will that give the right balance of reducing locks as the cache is populated but still make the read/write operations thread safe?

Reply:
> @davkean said it would reduce lock contention if we always look for an existing value and then lock before writing.

I never said this with `Dictionary<TKey, TValue>`. :)

You don't need to mix ConcurrentDictionary and the lock usage, use .GetOrAdd instead. However, I would still be interested in the contention this introduces, and you need to think about what happens when the factories overlap and one of the values is thrown away.